### PR TITLE
Support MySQL and MSSqlServer with Formula Engine

### PR DESCRIPTION
**Pull request by steventamm**
Mysql Changes:
  Use decimal(52,18) precision with Mysql
  Fix boolean output
  Round off milliseconds when adding numbers to dates
  Use TIMESTAMPDIFF not unix_timestamp for comparing dates
  Use DATE_ADD instead of +-, Concat, not ||
  Remove trailing zeros to allow easier visual comparison
  Fix AddMonths & Power
  Trying to use time value natively in mysql, but may give up
  Fix REGEXP in mysql needing backslash escaping
  Support rudimentary FormatCurrency in Mysql
  Bind values in mysql tester because literal strings remove control characters

MSSQLServer Changes:
  Note: MAX/MIN aren't supported on mssql
  Have an LPAD/RPAD implementation for tsql
  Add in a test for LEFT() and fix postgres failure
  Parameterize time math
  Put oracle datetime guard back to previous value
  Fix origin date time for tsql.
  Ignore arithmetic overflow on large date adds and comparisons for big numbers
  Use java.sql.time in mssqlcontainertester
  Use TransactSQL functions for date math with special values
  Use Decimal(38,18) from float arithmetic functions to handle type mismatches, and (38,10) from power/exp
  Have Today() parameterized as well
  Use VARCHAR not CHAR for conversion to prevent extra spaces
  Optimize mssql which prevents CASE WHEN .. THEN NULL ELSE NULL END
  Don't bind values in mssql tester because numbers are horked

FormulaSqlHooks:
  Move into separate classes for each sql style …
  Have add days to date be in SqlHooks to improve readability
  Have power/exponent be in SqlHooks to allow easier overriding of casting
   Make the TZ conversion for dateTime -> date more parameterizable since
  FormulaSqlHooks: Parameterize truncation, which changes mysql goldfiles
  Have addMonths take the values because they're in the reverse order in mssql
  Parameterize making strings case sensitive for comparison

General:
  FormulaImpl has a new parameter to handle substring difference in mssql
  Clean up pom file dependencies
  Also move most of the ignore-test stuff out of the testers into the xml file
  Make AbstractDbTester to reuse code between embedded and testcontainers
  Use of BigDecimal.stripTrailingZeros is shared everywhere
  Better commentary on connection reuse
  Add testWeekday back in
  Test Min/Max in formulatests.xml to validate greatest/least
  Test negative values with formatcurrency and simplify formatcurrency hooks logic. 
  Make timezone offset/names more parameterized
  Small change psql for formatting of timestamp to use concat, not ||

### CODE REVIEW

File: api/src/main/java/com/force/formula/sql/FormulaSqlStyle.java [hunk 1]
```diff
@@ -6,25 +6,50 @@
...
+	 * @return whether or not to default to transactsql (sybase/ms sql server) style of sql.
+	 */
+	default boolean isTransactSqlStyle() {
+		return false;
+	}	
+	
+	/**
+	 * @return the function name for taking a substring.
+	 */
+	default String getSubstringFunction() {
```
Comment (steventamm) [inline api/src/main/java/com/force/formula/sql/FormulaSqlStyle.java:52]: I'm putting this here because it's referenced in the api module as part for

File: impl/src/main/java/com/force/formula/commands/FunctionAddMonths.java [hunk 2]
```diff
@@ -32,7 +32,7 @@ public FormulaCommand getCommand(FormulaAST node, FormulaContext context) {
 
     @Override
     public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, String[] guards, TableAliasRegistry registry) {
-        String sql = String.format(getSqlHooks(context).sqlAddMonths(), args[0], args[1]);
+        String sql = getSqlHooks(context).sqlAddMonths(args[0], args[1]);
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionAddMonths.java:35]: SQLServer changes the order, so we can't use String.format

File: impl/src/main/java/com/force/formula/commands/FunctionDate.java [hunk 3]
```diff
@@ -44,7 +45,14 @@ public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, St
         String year = (yearValue != BAD_VALUE) ? String.valueOf(yearValue) : String.format(getSqlHooks(context).sqlToChar(), "FLOOR(" + args[0] + ")");
         String month = (monthValue != BAD_VALUE) ? String.valueOf(monthValue) : String.format(getSqlHooks(context).sqlToChar(), "FLOOR(" + args[1] + ")");
         String day = (dayValue != BAD_VALUE) ? String.valueOf(dayValue) : String.format(getSqlHooks(context).sqlToChar(), "FLOOR(" + args[2] + ")");
-        String date = "TO_DATE(" + year + " || '-' || " + month + " || '-' || " + day + ", 'YYYY-MM-DD')";
+        String date;
+        if (hooks.isMysqlStyle()) {
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionDate.java:49]: I didn't put this in the SqlHooks yet.  I probably should

File: impl/src/main/java/com/force/formula/commands/FunctionExponent.java [hunk 4]
```diff
@@ -38,13 +37,9 @@ protected BigDecimal execute(BigDecimal arg) {
...
     public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, String[] guards) {
-        if (context.getSqlStyle() != null && context.getSqlStyle().isPostgresStyle()) {
-            // tests showed double precision was only 2.5% faster than numeric (i.e. the rest of
-            // the query processing machinery dominates, so go for max precision
-            return new SQLPair("EXP(" + args[0] + "::numeric(40,20))", guards[0]);
-        } else {
-            return new SQLPair("EXP(" + args[0] + ")", guards[0]);
-        }
+    	FormulaSqlHooks hooks = (FormulaSqlHooks) context.getSqlStyle();
+    	String sql = hooks.sqlExponent(args[0]);
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionExponent.java:41]: Exponent in sqlserver was complicated enough for me to move it to hooks to allow it to be overridden

File: impl/src/main/java/com/force/formula/commands/FunctionFind.java [hunk 5]
```diff
@@ -30,14 +30,15 @@ public FormulaCommand getCommand(FormulaAST node, FormulaContext context) {
...
         StringBuilder sql = new StringBuilder();
 
+        FormulaSqlHooks hooks = getSqlHooks(context); 
         if (args.length == 3) {
         	// With postgres, you need to do a bunch of stuff to get around the lack of a built-in for the offset.
-            String startPosition = getSqlHooks(context).sqlNvl() + "(" + args[2] + ", 1)";
-	        String instr = getSqlHooks(context).sqlInstr3(args[1], args[0], "GREATEST("+startPosition+",1)");
-	        sql.append(String.format(getSqlHooks(context).sqlNvl() + "(" + instr + ", 0)", args[1], args[0], startPosition));
+            String startPosition = hooks.sqlNvl() + "(" + args[2] + ", 1)";
+	        String instr = hooks.sqlInstr3(args[1], args[0], hooks.sqlGreatest(startPosition, "1"));
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionFind.java:37]: SqlServer doesn't have GREATEST/LEAST until MS SqlServer 2022.... so I had to parameterize it

File: impl/src/main/java/com/force/formula/commands/FunctionLpad.java [hunk 6]
```diff
@@ -34,8 +34,9 @@ public FormulaCommand getCommand(FormulaAST node, FormulaContext context) {
 
     @Override
     public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, String[] guards, TableAliasRegistry registry) {
-        String sql = node.getNumberOfChildren() == 3 ? "LPAD(" + args[0] + ", GREATEST(" + getSqlHooks(context).sqlRoundScaleArg(args[1]) + ", 0), " + args[2]+ ")"
-            : "LPAD(" + args[0] + ", GREATEST(" + getSqlHooks(context).sqlRoundScaleArg(args[1]) + ", 0))";
+        FormulaSqlHooks hooks = getSqlHooks(context);
+        String amount = hooks.sqlEnsurePositive(hooks.sqlRoundScaleArg(args[1]));  // prevent negative numbers
+        String sql = hooks.sqlLpad(args[0], amount, node.getNumberOfChildren() == 3 ? args[2] : null);
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionLpad.java:39]: SqlServer doesn't have greatest... nor LPAD....  so I parameterized it

File: impl/src/main/java/com/force/formula/commands/FunctionMod.java [hunk 7]
```diff
@@ -38,7 +38,12 @@ protected BigDecimal execute(BigDecimal lhs, BigDecimal rhs) {
 
     @Override
     public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, String[] guards) {
-        String sql = "MOD(" + args[0] + ", " + args[1] + ")";
+        String sql;
+        if (context.getSqlStyle().isTransactSqlStyle()) {
+        	sql = "(" + args[0] + " % " + args[1] + ")";
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionMod.java:43]: SQLServer only supports % operator

File: impl/src/main/java/com/force/formula/commands/FunctionNaturalLog.java [hunk 8]
```diff
@@ -34,7 +33,13 @@ protected BigDecimal execute(BigDecimal arg) {
 
     @Override
     public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, String[] guards) {
-        String sql = "LN(" + args[0] + ")";
+        String sql;
+    	FormulaSqlHooks hooks = (FormulaSqlHooks) context.getSqlStyle();
+        if (hooks.isTransactSqlStyle()) {
+        	sql = String.format(hooks.sqlToNumber(),"LOG(" + args[0] + ")");
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionNaturalLog.java:39]: The SqlToNumber is needed for float -> decimal.

File: impl/src/main/java/com/force/formula/commands/FunctionText.java [hunk 9]
```diff
@@ -161,29 +161,27 @@ public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, St
...
-                    sqlBuilder.append(", 'YYYY-MM-DD'");
-                } else if (clazz == FormulaDateTime.class) {
-                    sqlBuilder.append(", 'YYYY-MM-DD HH24:MI:SS'");
-                }
-
-                sqlBuilder.append(")");
-
+            	String toChar;
+            	if (clazz == Date.class || parentType == Date.class) {
+            		toChar = String.format(hooks.sqlToCharDate(), args[0]);
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionText.java:172]: All the date stuff is different in mysql & sqlserver, so it's in the hooks.

File: impl/src/main/java/com/force/formula/commands/FunctionTrunc.java [hunk 10]
```diff
@@ -40,7 +40,7 @@ protected BigDecimal execute(BigDecimal lhs, BigDecimal rhs) {
     @Override
     public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, String[] guards) {
     	FormulaSqlHooks hooks = (FormulaSqlHooks) context.getSqlStyle();
-        String sql = "TRUNC(" + args[0] + ", " + hooks.sqlRoundScaleArg(args[1]) + ")";
+        String sql = hooks.sqlTrunc(args[0],hooks.sqlRoundScaleArg(args[1]));
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/FunctionTrunc.java:43]: SqlServer uses a third argument for ROUND to truncate (i.e. `TRUNC(foo,0)` is `ROUND(foo,0,1)`)

File: impl/src/main/java/com/force/formula/commands/OperatorAddOrSubtract.java [hunk 11]
```diff
@@ -177,32 +177,24 @@ public SQLPair getSQL(FormulaAST node, FormulaContext context, String[] args, St
...
-                    }
+                    sql = String.format(getSqlHooks(context).sqlSubtractTwoTimestamps(), lhsValue, rhsValue);
+                } else if (isDateTimeAndNumberOperation(lhsDataType, rhsDataType) && !hooks.isOracleStyle()) {
+                	sql = hooks.sqlAddDaysToDate(lhsValue, lhsDataType, rhsValue, rhsDataType, !"-".equals(operator));
                 } else if ("||".equals(operator)) {
                     sql = "(" + String.format(getSqlHooks(context).sqlConcat(false), args[0], args[1]) + ")";
                 } else {
                     sql = "(" + lhsValue + operator + rhsValue + ")";
                 }
-            } else {
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/OperatorAddOrSubtract.java:204]: Moved above because oracle is easy.

File: impl/src/main/java/com/force/formula/commands/OperatorEquality.java [hunk 12]
```diff
@@ -217,19 +217,26 @@ public static String compareBulk(FormulaContext context, Object lhs, int lhsType
...
             // the other side unless both are null. Cool eh? Note this relies on (null || 'x')
-            // being 'x'.  In postgres, it relied on concat(...)
+            // being 'x'.  In postgres, it relies on concat(...)
         	FormulaSqlHooks sqlHooks = (FormulaSqlHooks) context.getSqlStyle();
             Object saveRhs = rhs;
             
             if (rhsType != FormulaTokenTypes.STRING_LITERAL) {
                 rhs = sqlHooks.sqlNvl() + "(" + rhs + ", " + String.format(sqlHooks.sqlConcat(false), lhs,  "'x'") + ")";
-            } else if ("''".equals(rhs)) {
+            } else if ("''".equals(rhs) || ("NULL".equals(rhs) && sqlHooks.isPostgresStyle())) {  // see ConstantString for postgres
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/commands/OperatorEquality.java:226]: ConstantString replaces `"''"` with `NULL` for Postgres only to match oracle semantics.  I didn't do this for mysql/mssql... yet.  But if you do, it breaks formulas that compare `text = ''` or `text <> ''`.  This fixes it by treating NULL string like '' for Postgres.  If the change in ConstantString is done in other languages, this would have to change.

File: impl/src/main/java/com/force/formula/impl/BaseFormulaInfoImpl.java [hunk 13]
```diff
@@ -793,6 +794,13 @@ public boolean hasAIPredictionFieldReference() {
     }
 
     @Override
+	public boolean hasFormatCurrencyCommand() {
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/BaseFormulaInfoImpl.java:797]: Now the FormatCurrency's in the open-source version, this came back for testing.

File: impl/src/main/java/com/force/formula/impl/FormulaUtils.java [hunk 14]
```diff
@@ -586,62 +584,6 @@ public static boolean isTypeSobjectRow(FormulaDataType type) {
...
-        return FormulaTypeUtils.isTypeText(type);
-    }
-
-    /**
-     * Same as isTypeText, but in cases where we think supporting an ID is ugly.
-     * @param type the type to test that may be an ID
-     * @return  is the type text (or ID)
-     * @deprecated use FormulaTypeUtils
-     */
-    @Deprecated
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/FormulaUtils.java:610]: Removed a bunch of deprecated methods.

File: impl/src/main/java/com/force/formula/impl/sql/FormulaDefaultSqlStyle.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+package com.force.formula.impl.sql;
+
+import com.force.formula.impl.FormulaSqlHooks;
+
+/**
+ * @author stamm
+ *
+ */
+public final class FormulaDefaultSqlStyle {
+	public static final FormulaSqlHooks POSTGRES = new FormulaPostgreSQLHooks() {};
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaDefaultSqlStyle.java:13]: I changed the enum to constants instead, to avoid doing enum testing and make the logic easier to read with OOP.  Now there's a lot fewer if/then/elses.

File: impl/src/main/java/com/force/formula/impl/sql/FormulaMySQLHooks.java [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+    	if (pad == null) {
+        	return "RPAD(" + str + ", " + amount + ", ' ')";    		
+    	} else {
+    		return FormulaSqlHooks.super.sqlRpad(str, amount, pad);
+    	}
+    }
+	
+	@Override
+    default Object sqlMakeCaseSensitiveForComparison(Object str) {
+		return "binary " + str;
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaMySQLHooks.java:295]: Mysql and mysql are case insensitive by default, and making them case sensitive is difficult.  So the usual thing is to make the particular comparison case sensitive by adding COLLATE.  But mysql requires binary

File: impl/src/main/java/com/force/formula/impl/sql/FormulaPostgreSQLHooks.java [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+     * In oracle and mysql, it's INSTR.  Use binary for INSTR to keep it case sensitive
+     * @param strArg the value of the string to search in
+     * @param substrArg the value of the substring to search
+     * @param startLocation the value of the start location to use as the offset (1-indexed)
+     */
+	@Override
+    default String sqlInstr3(String strArg, String substrArg, String startLocation) {
+		// This is unfortunate, as it has to reevaluate the subexpression twice in order to return 0 for not found.
+		// If your postgresql has a better version of this, please use it instead.
+		return String.format("CASE WHEN COALESCE(STRPOS(SUBSTR(%s,%s::integer),%s),0) > 0 THEN STRPOS(SUBSTR(%s,%s::integer),%s) + %s - 1 ELSE 0 END", strArg, startLocation, substrArg, strArg, startLocation, substrArg, startLocation);
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaPostgreSQLHooks.java:212]: Native Postgres doesn't have this.  Sigh.

File: impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+    default String sqlInstr3(String strArg, String substrArg, String startLocation) {
+		return String.format("CHARINDEX(%s COLLATE Latin1_General_CS_AS, %s, %s )", substrArg, strArg, startLocation);
+    }
+	
+	@Override
+    default String sqlIsNumber() {
+		// This function is not very good, as it does a simple character test and 
+		// allows - and + in all positions.  MSSqlServer doesn't support regex natively,
+		// So this is about all we can do without doing NOT LIKE '%[^0-9.+-]'...
+    	return "ISNUMERIC(%s)=1";
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java:93]: ISNUMERIC is considered harmful, but since mssql doesn't have native regex... there's not much else that can be done.

File: impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+    
+	// See https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15
+    @Override
+    default String sqlToCharTimestamp() {
+        return "CONVERT(VARCHAR,%s,120)";  // 120 = "yyyy-mm-dd hh:mi:ss (24h)"
+     }
+    
+    @Override
+    default String sqlToCharDate() {
+        return "CONVERT(VARCHAR,%s,105)";  // 105 = "dd-mm-yyyy"
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java:121]: The magic numbers are documented above

File: impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+        String sql = "CAST(POWER(" + args[0] + ", " + args[1] + ") AS DECIMAL(38,10))";
+        String guard = SQLPair.generateGuard(guards, "ROUND(" + args[1] + ",0,1)<>" + args[1] +
+	            " OR(" + args[0] + "<>0 AND LOG10(ABS(" + args[0] + "))*" + args[1] + ">38)");    	
+        return new SQLPair(sql, guard);
+    }
+	
+	// MSSqlServer doesn't support greatest or least until 2022.  So we do (a+b+ABS(a-b)/2)
+	@Override
+	default String sqlGreatest(String arg1, String arg2) {
+		return "(" + arg1 + "+" + arg2 + "+ABS(" + arg1 + "-" + arg2 + "))/2";
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java:212]: This is smelly enough, but was the best thing I found online.  I didn't fix MAX()/MIN() in mssql because it was so awful.  Once azure's on 2022, it should be moot.

File: impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java [hunk 21]
```diff
@@ -0,0 +1,275 @@
...
+
+	@Override
+    default String sqlEnsurePositive(String argument) {
+		// SQL Server doesn't have greatest.  So do (a+abs(a))/2
+    	return "((" + argument + " + ABS(" + argument + "))/2)";
+    }
+
+
+	@Override
+	default String sqlLpad(String str, String amount, String pad) {
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java:224]: Implementation of LPAD.  Much more difficult than RPAD, as the comment states

File: impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+
+	
+    /**
+     * @param scale the number of digits to the right of the radix
+     * @return the SQL string to use to in TO_CHAR for the given scale.  This is used by 
+     * {@link #getCurrencyFormat(String, String, boolean)}
+     */
+	@Override
+    default StringBuilder getCurrencyMask(int scale) {
+        StringBuilder mask = new StringBuilder(40).append("'###,###,###,###,##0");
```
Comment (steventamm) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java:260]: Use excel format style in mssql

File: impl/src/test/java/com/force/formula/template/commands/TemplateFunctionsTest.java [hunk 23]
```diff
@@ -352,5 +355,62 @@ public void testFormatCurrency() throws Exception {
         }
     }
 
-    
+    /**
+     * Test parsing with other parse options to make sure it returns the right errors
+     * @throws Exception
+     */
+    public void testParseTemplateWithBoth() throws Exception {
```
Comment (steventamm) [inline impl/src/test/java/com/force/formula/template/commands/TemplateFunctionsTest.java:362]: Bump up code coverage by testing antlr2 & antlr4.
@armanshan12

File: mysql-test/src/test/java/com/force/formula/impl/TestMariaDBExtendedFormulas.java [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+
+import junit.framework.TestSuite;
+
+/**
+ * Contains non-extended tests for formulatests.xml in MariaDB
+ * @author stamm
+ * @since 0.2
+ */
+@RunWith(AllTests.class)
+@Ignore
```
Comment (steventamm) [inline mysql-test/src/test/java/com/force/formula/impl/TestMariaDBExtendedFormulas.java:27]: All MariaDB tests are ignored for now because some of them break on corner cases and it appears to be really slow to run in test containers.

File: sqlserver-test/src/test/java/com/force/formula/impl/FormulaMsSqlServerTests.java [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+	protected static class MSSQLFormulaTest extends FormulaGenericTest {
+		
+		public MSSQLFormulaTest(FormulaTestCaseInfo testCase, String name, boolean positive,
+				FormulaGenericTests suite) {
+			super(testCase, name, positive, suite);
+		}
+		
+		@Override
+		protected String getSqlErrorMessageResult(String errorViaSql, String viaFormula) {
+			// SqlServer can only handle 38 digits of precision, and because of the use of
```
Comment (steventamm) [inline sqlserver-test/src/test/java/com/force/formula/impl/FormulaMsSqlServerTests.java:85]: There were a lot of these in the goldfields with things like EXP(100) failing or doing dates beyond 2100.

File: test-utils/src/main/java/com/force/formula/impl/FormulaGenericTests.java [hunk 26]
```diff
@@ -145,7 +154,7 @@ protected boolean shouldTestJavascript() {
 		 * @return whether for a give test, the values from the SL engine should be compared
 		 */
 		protected boolean shouldCompareSql() {
-			return shouldTestSql() && !getTestCaseInfo().ignoreSql();
+			return shouldTestSql() && null == getTestCaseInfo().whyIgnoreSql(((FormulaGenericTests)getSuite()).getDbTypeName());
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/impl/FormulaGenericTests.java:157]: With whyIgnoreSql you have to say which DB you want now

File: test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java [hunk 27]
```diff
@@ -0,0 +1,457 @@
...
+ * A default implementation of DbTester of formulas that simplifies some stuff around
+ * creating the test case SQL statements.
+ * 
+ * Instead of writing rows to the database, it puts all of the "passed in
+ * values" in a subquery with alias "c", casting the null values to the right types.
+ * 
+ * @author stamm
+ * @since 0.2
+ */
+public abstract class AbstractDbTester implements DbTester {
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java:29]: This abstracts the EmbeddedPostgres/TestContainer stuff from the nuts and bolts of generating the SQL

File: test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java [hunk 28]
```diff
@@ -0,0 +1,457 @@
...
+	 */
+	protected String fixSqlFormat(FormulaContext context, String columnSql, Formula formula) {
+		MockFormulaDataType returnType = (MockFormulaDataType) formula.getDataType();
+		switch (returnType) {
+		case DATETIME:
+			return convertToDateTime("(" + columnSql + ")");
+		case PERCENT:
+			return "(" + columnSql + ")/100"; // The format below moves the decimal point over
+		case BOOLEAN:
+			if (!columnSql.endsWith(" THEN '1' ELSE '0' END")) {  // TODO: This shouldn't work.
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java:155]: There are some tests that return a boolean that seem to be double converted.  So I put this in to stop that

File: test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java [hunk 29]
```diff
@@ -0,0 +1,457 @@
...
+	 *                       fields
+	 * @param values         the values to mock
+	 * @return a subquery to evaluate field references
+	 */
+	protected CharSequence makeRowValueSubquery(FormulaRuntimeContext formulaContext, Map<String, Object> values) {
+		MapFormulaContext mfc = (MapFormulaContext) formulaContext;
+
+		// Construct subquery
+		StringBuilder sub = new StringBuilder();
+		sub.append(" FROM (SELECT 1 as fake");
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java:198]: MSSql requires an alias on all literals in a subquery.

File: test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java [hunk 30]
```diff
@@ -0,0 +1,457 @@
...
+		sub.append(") c");
+		return sub;
+	}
+	
+	
+	/**
+	 * @return true if we should use binding variables instead of literals. 
+	 */
+	protected boolean useBinds() {
+		return false;
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java:226]: Oracle and Postgres use literals because they did before.  MySql uses binds because literal strings are broken otherwise.  MSSql uses literals because the JDBC driver ignores whatever scale is on the big decimal and uses whatever it thinks it should be.

File: test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java [hunk 31]
```diff
@@ -0,0 +1,457 @@
...
+			if (bigDecimal == null)
+				return null;
+			// Currency and Percent have special formatting.
+			return String.valueOf(FormulaI18nUtils.formatResult(formulaContext, formulaContext.getFormulaReturnType(), bigDecimal));
+		case INTEGER:
+		case DOUBLE:
+			BigDecimal number = rset.getBigDecimal(1);
+			if (number == null)
+				return null;
+			return number.stripTrailingZeros().toPlainString(); // Strip trailing zeros because that's driver specific.
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/sql/AbstractDbTester.java:347]: Specifically, the Postgres and mysql driver add a bunch of zeros

File: test-utils/src/main/java/com/force/formula/sql/PostgreSQLContainerTester.java [hunk 32]
```diff
@@ -8,17 +8,22 @@
 import org.testcontainers.containers.PostgreSQLContainer;
 
 /**
- * Postgres tester that uses a container. 
+ * PostgreSQL tester that uses a container.
```
Comment (steventamm) [inline test-utils/src/main/java/com/force/formula/sql/PostgreSQLContainerTester.java:11]: Match the name of the testconatiner

File: impl/src/test/java/com/force/formula/template/commands/TemplateFunctionsTest.java (hunk 23 above)
Comment (armanshan12) [reply impl/src/test/java/com/force/formula/template/commands/TemplateFunctionsTest.java:362]: LGTM. I'm noticing that this class is testing more than templates, so the class name might be misleading.

File: impl/src/main/java/com/force/formula/impl/sql/FormulaMySQLHooks.java [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+import java.lang.reflect.Type;
+import java.util.*;
+
+import com.force.formula.FormulaDateTime;
+import com.force.formula.FormulaTime;
+import com.force.formula.impl.FormulaSqlHooks;
+import com.force.formula.sql.SQLPair;
+
+/**
+ * Implementation of FormulaSqlHooks for Oracle-style DBs
```
Comment (bmazyan) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaMySQLHooks.java:15]: minor, doc - should this be "Implementation of FormulaSqlHooks for MySQL-style DBs" instead of "...Oracle-style DBs" ?

File: impl/src/main/java/com/force/formula/impl/sql/FormulaPostgreSQLHooks.java [hunk 34]
```diff
@@ -0,0 +1,249 @@
...
+package com.force.formula.impl.sql;
+
+import java.lang.reflect.Type;
+import java.util.Date;
+
+import com.force.formula.FormulaDateTime;
+import com.force.formula.impl.FormulaSqlHooks;
+
+/**
+ * Implementation of FormulaSqlHooks for Oracle-style DBs
```
Comment (bmazyan) [inline impl/src/main/java/com/force/formula/impl/sql/FormulaPostgreSQLHooks.java:13]: minor, doc - should this be "Implementation of FormulaSqlHooks for PostgreSQL-style DBs"?

File: impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java (hunk 20 above)
Comment (bmazyan) [reply impl/src/main/java/com/force/formula/impl/sql/FormulaTransactSQLHooks.java:212]: minor, doc: "So we do (a+b+ABS(a-b)/2)" - should be (a+b+ABS(a-b))/2

**Review by bmazyan — APPROVED** [review]

**Review by bmazyan — APPROVED** [review]
